README.md
Chrome-Addon-Fork
=================

g1plus Addon für Chrome Download: http://g1plus.x10.mx/

Beschreibung
============

GameOne.de ist ein Portal rund um Videospiele, welches ständig neue Inhalte im Videoformat veröffentlicht. Jedoch nur als Stream - ein Download dieser Videos wird nicht angeboten, was dem Nutzer eine Reihe von Dingen erschwert/verbietet:

* Videos auf anderen Geräten mit nur unzureichender/fehlender Flash-Unterstützung (z.B. PS3) sehen
* Videos offline sehen (z.B. auf Zugfahrten oder auf Reisen)
* Videos die der Altersbeschränkung unterliegen auch innerhalb der Sperrzeiten sehen (einige Videos sind nur in einem bestimmten Zeitraum verfügbar, welcher jedoch nicht für alle günstig ist)
* Videos durchgängig in der höchsten Qualität sehen ungeachtet der Bandbreite (bei schwacher Bandbreite passt der verwendete Player auf der Seite adaptiv die Qualität an, was dazu führt, dass sich ständig hohe und niedrige Qualität abwechseln)

Disclaimer
==========

Mit dem AddOn sollen diese Probleme gelöst werden, indem Download-Links unterhalb der entsprechenden Streaming-Inhalte angezeigt werden. Es liegt aber nicht in unserer Absicht damit die Verteilung der Inhalte auf anderen Video-Plattformen, wie YouTube, zu begünstigen. Ebenfalls liegt es uns fern die Werbeeinnahmen von GameOne durch verlorene Seitenaufrufe zu schmälern: die Download-Links werden in die Seite integriert und somit ist es immer noch notwendig die entsprechenden Beiträge aufzurufen. Ebenfalls wird die Altersbeschränkung nicht komplett aufgehoben, wenn auch durch eine schwächere Altersabfrage ersetzt. Dieser, technisch gesehen, nicht notwendige Schritt wird auch in kommenden Versionen nicht verschwinden.


CHANGELOG.md
ChangeLog
=========

### V0.4
* Support für den neuen Player
* Die Chrome Erweiterung kann sich automatisch aktualisieren (Auto-Update)

### V0.3.6
* Online update der gesperrten Videodatenbank

### V0.3.5
* Anzeigen der Länge des Videos
* Anzeigen des Vorschaubildes von Videos, selbst bei Aktiviertem Flashblock 

### V0.3.4
* Entsperren von Inhalten, aber nur basieren auf offline Cache, kein on-the-fly Update

### V0.3.3
* Fork vom Firefox Addon 
* generiert Downlodlink aber kein entsperren von Inhalten


lib/main.js
/* Import
 * ====== */

var data = self.data;

/* Const
 * ===== */

var cacheMirrors = ['http://pastebin.com/raw.php?i=tVLCAjZc',
                    'http://g1plus.x10.mx/cache.json'];

/* Methoden
 * ======== */

/**
 * Synchronsieren des lokalen cache mit der letzten Version
 * auf pastebin oder eines Fallback-Mirrors, wenn neuer.
 *
 * @param onSync  Callback der nach dem Sync-Vorgang aufgerufen wird.
 *                Der Status des Sync-Versuches wird übergeben.
 */
function syncCache(mirror, onSync) {
    var req = request.Request({
        url: cacheMirrors[mirror],
        onComplete: function(response) {
            if(response.status == 200) {
                var json = JSON.parse(response.text);
                if(storage.cache['update']) {
                    if(Date.parse(json['update']) > Date.parse(storage.cache['update'])) {
                        storage.cache = json;
                    }
                } else {
                    storage.cache = json;
                }
                onSync(response.status);
            } else {
                if(cacheMirrors[mirror + 1]) {
                    syncCache(mirror + 1, onSync);
                } else {
                    onSync(response.status);
                }
            }
        }
    });
    req.get();
}

/**
 * Filter die Content-Elemente aus einer Seite.
 *
 * @param src  Quelltext der Seite.
 *
 * @return Die Content-Elemente reduziert auf ihre Referenzen.
 */
function getContentElements(src) {
    var re = /#gallery_\d*|video_meta-\w*|<embed(.*)/g;

    var result = new Array();

    while((match = re.exec(src)) != null) {
        if(match[0].indexOf('embed') > -1) {
            result.push(match[0].split('"')[1]);
        } else if(match[0].indexOf('video_meta-') > -1) {
            result.push(match[0].split('-')[1]);
        } else {
            result.push(match[0].split('#')[1]);
        }
    }

    return result;
}

/**
 * Erzeugt ein Cache-Objekt aus der Differenz zweier HTML-Seiten.
 *
 * @param pre  Die Seite mit gesperrten Inhalten.
 * @param post Die Seite mit 18er-Inhalten.
 *
 * @return Cache-Objekt
 */
function diff(pre, post) {
    var result = {};

    var re = /dummy_agerated.jpg/g;

    if(!re.test(post)) {
        var post_elems = getContentElements(post);
        var pre_elems = getContentElements(pre);

        for(i in pre_elems) {
            var index = post_elems.indexOf(pre_elems[i]);
            if(index > -1) {
                post_elems.splice(index, 1);
            }
        }

        var dummy_count = 0;
        var match = re.exec(pre);

        while(match) {
            ++dummy_count;
            var src = post_elems.shift();
            result[String(dummy_count)] = new Array(src);
            match = re.exec(pre);
        }

        if(dummy_count < post_elems) {
            result[String(dummy_count)].concat(post_elems);
        }
    }

    return result;
}

/**
 * Legt ein Cache-Objekt in den lokalen Storage.
 *
 * @param id  id der Seite
 * @param page  page der Seite
 * @param cache Cache-Objekt
 */
function addToCache(id, page, cache) {
    if(cache['1'] && cache['1'].length > 0) {
        if(!storage.cache[id]) {
            storage.cache[id] = {};
        }
        if(!storage.cache[id][page]) {
            storage.cache[id][page] = cache;
        }
    }
}

/**
 * Überprüft auf ein Update des AddOns und führt folgende Aktionen
 * aus:
 *  - aufrufen des Changelogs
 *  - wenn notwendig aktualisieren des lokalen Caches mit dem beim
 *    Update mitgelieferten
 */
function checkUpdate() {
    if(!storage.firstrun) {
        tabs.open('http://g1plus.x10.mx/information');
        storage.firstrun = true;
    }

    if(storage.version != self.version) {
        tabs.open('http://g1plus.x10.mx/changelog');
        storage.version = self.version;

        var cacheJson = JSON.parse(data.load('cache.json'));
        if(!storage.cache) {
            storage.cache = cacheJson;
        } else {
            // Update des Caches erzwingen, selbst wenn älter um einen fehlerhaften Cache zu überschreiben (z.B. fehlendes update-property *duh*)
            storage.cache = cacheJson;
        }
    }
}

data/augmentpage.js
/* Workarounds
 * ========== */

/** Bugfix for Webkit/Safari/Chrome with jQuery >= 1.7 for xml namespace $.find()
 * http://stackoverflow.com/questions/853740/jquery-xml-parsing-with-namespaces
 * http://www.steveworkman.com/html5-2/javascript/2011/improving-javascript-xml-node-finding-performance-by-2000/
 */
$.fn.filterNode = function(name) {
	return this.find('*').filter(function() {
		return this.nodeName === name;
	});
};

/* Funktionen
 * ========== */

/**
 * Erzeugt einen Container für Download-Links.
 *
 * @param id   Die id die des Containers
 *
 * @return Der Container
 */
function createDownloadContainer(id)
{
    var downloads = document.createElement('div');
    downloads.setAttribute('id', id)
    downloads.setAttribute('class', 'downloads g1plus');

    var heading = document.createElement('h4');
    heading.textContent = 'Downloads';
    downloads.appendChild(heading);

    return downloads;
}

function createWarning(msg)
{
    var warning = document.createElement('div');
    warning.setAttribute('class', 'warn_text');
    $(warning).html(msg);

    return warning;
}

/**
 * Erzeugt einen simplen Link.
 *
 * @param url   Url auf die der Link zeigt
 * @param text  Text des Links
 *
 * @return Das Link-Element
 */
function createDownloadLink(url, text)
{
    var downlink = document.createElement('a');
    downlink.setAttribute('href', url);
    downlink.textContent = text;

    return downlink;
}

/**
 * Erstellt einen Player, der dem GameOne-Player entspricht.
 *
 * @param src  Quelle des anzuzeigenden Videos.
 */
function createPlayer(src)
{
    var player = document.createElement('embed');
    player.setAttribute('width', 566);
    player.setAttribute('height', 424);
    player.setAttribute('flashvars', 'configParams');
    player.setAttribute('menu', 'false');
    player.setAttribute('swliveconnect', 'true');
    player.setAttribute('allowscriptaccess', 'always');
    player.setAttribute('enablejavascript', 'true');
    player.setAttribute('allowfullscreen', 'true');
    player.setAttribute('quality', 'high');
    player.setAttribute('name', 'embeddedPlayer');
    player.setAttribute('id', 'embeddedPlayer');
    player.setAttribute('src', src);
    player.setAttribute('type', 'application/x-shockwave-flash');
    return player;
}

/**
 * Holt die zugehörigen Downloads des Owner-Objects und fügt sie in einer
 * Download-Box an.
 */
function getDownloads()
{
    var src = this.getAttribute('src');
    var id = src.split('-').pop();
    this.parentNode.appendChild(createDownloadContainer('downloads_' + id));
    $.get('http://gameone.de/api/mrss/' + src, function(data) { response_mrss(data, id) } );
}

/**
 * Füllt ein select-Element mit option-Einträgen beginnend bei min bis max
 * aufsteigend. Wird kein callback angegeben entsprech die Einträge den
 * jeweiligen Zahlen.
 *
 * @param select    Das zu befüllende select-Element
 * @param min       Startelement
 * @param max       Endelement
 * @param callback  Funktion die für jedes Element aufgerufen wird. Es wird die
 *                  jeweilige Ziffer übergeben. Rückgabewert muss ein dazu
 *                  korrespondierender String sein.
 */
function addOptions(select, min, max, callback)
{
	for(var i = min; i < max; ++i)
	{
		var option = document.createElement('option');
		if(!callback) {
			option.textContent = i;
		} else {
			option.textContent = callback(i);
		}
		select.appendChild(option);
	}
}

/**
 * Erzeugt eine Altersabfrage, die bei Eingabe eines Alters, welches der
 * Volljährigkeit entspricht, die Freigabe aller altersbeschränkten Inhalte
 * auslöst.
 *
 * @return Das Altersabfrage-Element
 */
function createAgeCheck()
{
	var agecheck = document.createElement('div');
	agecheck.setAttribute('class', 'agecheck g1plus');

	var info = document.createElement('p');
	info.textContent = 'Um Ab-18-Inhalte sehen zu können musst du dein Alter bestätigen:';
	agecheck.appendChild(info);

	var day = document.createElement('select');
	day.setAttribute('class', 'day');
	addOptions(day, 1, 32);
	agecheck.appendChild(day);

	var month = document.createElement('select');
	month.setAttribute('class', 'month');
	addOptions(month, 1, 13);
	agecheck.appendChild(month);

	var year = document.createElement('select');
	year.setAttribute('class', 'year');
	addOptions(year, 0, 100, function(i) {
		var today = new Date();
		return today.getFullYear() - i;
	});
	agecheck.appendChild(year);

	var ok = document.createElement('input');
	ok.setAttribute('type', 'submit');
	ok.setAttribute('value', 'Bestätigen');
	$(ok).click(function() {
		var year = parseInt($('select.year :selected', this.parentNode).text());
		var month = parseInt($('select.month :selected', this.parentNode).text());
		var day = parseInt($('select.day :selected', this.parentNode).text());
		var age = new Date(year, month - 1, day);
		var padded_age = new Date(age.getFullYear() + 18, age.getMonth(), age.getDate());
		var today = new Date();
		if((today.getTime() - padded_age.getTime()) >= 0) {
			var commentable_id = document.getElementById('commentable_id').getAttribute('value');
			$('div.agecheck').empty();
			$('div.agecheck').addClass('loading');
			self.port.emit('request_cache', {id: commentable_id, url: window.location.href});
		}
		return false;
	});
	agecheck.appendChild(ok);

	return agecheck;
}

/* Events
 * ==== */

/**
 * Behandeln der Rückgabe der mrss-API. Inhalte die von der flvgen-API geliefert
 * werden, werden gesondert behandelt. Doppelte Urls werden gefiltert.
 */
function response_mrss(data, id)
{
	var urls = new Array();
	$(data).filterNode('media:content').each(function() {
		var url = this.getAttribute('url');
		var callback = 'response_mediagen';

		if(url.indexOf('mediaGen.jhtml') != -1) {
			url = 'http://de.esperanto.mtvi.com/www/xml/flv/flvgen.jhtml?vid=' + url.split(':').pop();
			callback = 'response_flvgen';
		} else
			url = url.split('?')[0];

		if(urls.indexOf(url) == -1) {
			urls.push(url);

			if (callback == 'response_mediagen') {
			  $.get(url, function(data) { response_mediagen(data, id) } );
			} else {
			  $.get(url, function(data) { response_flvgen(data, id) } );
			}
		}
	});
}

/**
 * Behandeln von Inhalten, die über die mediagen-API geliefert werden (reguläre
 * Viedos sowie Gametrailers-Videos).
 */
function response_mediagen(data, id)
{
  var downloads = document.getElementById('downloads_' + id);
	var videos = [];

	$('rendition', data).each(function() {
		var v = {};
		v.width = this.getAttribute('width');
		v.height = this.getAttribute('height');
		v.bitrate = this.getAttribute('bitrate');
		v.mime = this.getAttribute('type').split('/').pop();
		if(this.textContent.indexOf('http') == -1) {
			v.url = this.textContent.trim().split('/riptide/').pop();
			v.url = 'http://cdn.riptide-mtvn.com/' + v.url;
		} else {
			v.url = this.textContent;
		}
		videos.push(v);
	});

	videos.sort(function(a, b) {
		return b.width - a.width;
	});

	videos.sort(function(a, b) {
		return b.bitrate - a.bitrate;
	});

	$(videos).each(function () {
		var downlink = createDownloadLink(this.url,
			this.mime + ' ' + this.width + 'x' + this.height + '@' + this.bitrate + 'kbps');
		downloads.appendChild(downlink);
	});
}


/**
 * Behandeln von Inhalten, die von der flvgen-API geliefert werden (TV-Folgen 1 - 150).
 */
function response_flvgen(data, id)
{
	var downloads = document.getElementById('downloads_' + id);
	items = [];

	$('src', data).each(function() {
			items.push($(this).text());
	});

	$(items).each(function() {
		var text = this.split('/').pop();
		text = text.split('.').shift();
		var downlink = createDownloadLink(this, text);
		downloads.appendChild(downlink);
	});

	var x = $('a', downloads);

	x.sort(function(a, b) {
		return b.textContent < a.textContent;
	});

	$('a', downloads).each(function() {
		downloads.removeChild(this);
	});

	$(x).each(function() {
		downloads.appendChild(this);
	});
}


/* Main
 * ==== */

// Downloads für alle Videos holen
$('div.player_swf embed').each(getDownloads);

/*
if($('img[src="/images/dummys/dummy_agerated.jpg"]').length == 0) {
    if(document.getElementById('commentable_id')) {
        var commentable_id = document.getElementById('commentable_id').getAttribute('value');
        self.port.emit('add_to_cache', {id: commentable_id, url: window.location.href});
    }
} else { // Altersbeschränkte Inhalte mit einer Altersfreigabe versehen
    $('img[src="/images/dummys/dummy_agerated.jpg"]').each(function(i) {
        $(this).replaceWith(createAgeCheck());
    });
}
*/


scripts/background_page.js
/* Const
 * ===== */

var cacheMirrors = ['http://pastebin.com/raw.php?i=tVLCAjZc',
                    'http://g1plus.x10.mx/cache.json'];

var informationPage = 'http://g1plus.x10.mx/blog/information/';

var backup_cache_url = chrome.extension.getURL("/data/backup_cache.json");
var online_cache = false;

/**
 * Altersfreigabe Daten laden die mit der Erweiterung als Backup/Fallback kommen
 * Werden direkt aus der Date geladen da der lokale Storage der von einer Extension
 * verändert werden kann auf 512 Einträge mit je 2KB begrenzt ist.
 * http://code.google.com/chrome/extensions/trunk/storage.html
 */
function loadBackupCache()
{
  var xhr = new XMLHttpRequest();
  xhr.open("GET", backup_cache_url, false);
  xhr.onreadystatechange = function()
  {
    if (xhr.readyState == 4) {
      if(localStorage.cache == undefined) {
        localStorage.cache = xhr.responseText;
      }
      else {
        // test if valid data present
        try {
          JSON.parse(localStorage.cache);
        } catch (e) {
          // no valid json in local cache override with backup
          localStorage.cache = xhr.responseText;
        }
      }
    }
  };

  try {
    xhr.send();
  } catch(e) {
    console.log('Couldn\'t load backup_cache');
  }
}


/* Methoden
 * ======== */

function loadOnlineCache()
{
  online_cache = JSON.parse(localStorage.cache);
}


/**
 * Synchronsieren des lokalen cache mit der letzten Version
 * auf pastebin oder eines Fallback-Mirrors, wenn neuer
 */
function updateCacheFromWeb(mirror) {
  var xhr = new XMLHttpRequest();
  xhr.open("GET", cacheMirrors[mirror], true);
  xhr.onerror = function() {
    if(cacheMirrors[mirror + 1]) {
        updateCache(mirror + 1);
      }
  }
  xhr.onreadystatechange = function() {
    if (xhr.readyState == this.DONE) {
      try {
        var json = JSON.parse(xhr.responseText);
        if(localStorage.cache == undefined) {
          localStorage.cache = xhr.responseText;
          loadOnlineCache();
        }
        else {
          var cache_old = JSON.parse(localStorage.cache);
          if (cache_old['update'] && Date.parse(json['update']) > Date.parse(cache_old['update'])) {
            localStorage.cache = xhr.responseText;
          }
          loadOnlineCache();
        }
      } catch(e) {
        console.log('Couldn\'t update online_cache');
      }
    }
  }
  xhr.send();
}


function storeData(id, data) {
  localStorage[id] = data;
}

function getData(id) {
  // try to load data from online_cache
  if (online_cache[id] == undefined) {
    return 0; // TODO try to update cache here?
  } else {
    var a = online_cache[id];
    return online_cache[id];
  }
}

function onInstall() {
  // showInformation
  chrome.tabs.create({url: informationPage});
}

function onUpdate() {
  // show Changelog
  // TODO
}

function getVersion() {
  var details = chrome.app.getDetails();
  return details.version;
}

// Check if the version has changed.
var currVersion = getVersion();
var prevVersion = localStorage['version']
if (currVersion != prevVersion) {
  // Check if we just installed this extension.
  if (typeof prevVersion == 'undefined') {
    onInstall();
  } else {
    onUpdate();
  }
  localStorage['version'] = currVersion;
}

loadBackupCache();
updateCacheFromWeb(0);

/* Message-Request Listener
 * ======================== */
chrome.extension.onRequest.addListener(
  function(request, sender, sendResponse) {
    if (request.func == "getData") {
      sendResponse({data: getData(request.id)});
    }
});

scripts/augment_page.js
/* Workarounds
 * ========== */

/** Bugfix for Webkit/Safari/Chrome with jQuery >= 1.7 for xml namespace $.find()
 * http://stackoverflow.com/questions/853740/jquery-xml-parsing-with-namespaces
 * http://www.steveworkman.com/html5-2/javascript/2011/improving-javascript-xml-node-finding-performance-by-2000/
 */
$.fn.filterNode = function(name) {
  return this.find('*').filter(function() {
    return this.nodeName === name;
  });
};

/* Funktionen
 * ========== */

/**
 * Erzeugt einen Container für Download-Links.
 *
 * @param id   Die id die des Containers
 *
 * @return Der Container
 */
function createDownloadContainer(id)
{
    var downloads = document.createElement('div');
    downloads.setAttribute('id', id)
    downloads.setAttribute('class', 'downloads g1plus');

    var heading = document.createElement('h4');
    heading.textContent = '» Downloads';
    downloads.appendChild(heading);
    downloads.appendChild(document.createElement('br'));

    $(heading).click(function() {
      $(downloads).find('a').slideToggle();
      if (downloads.style.backgroundImage == '') {
        downloads.style.backgroundImage = 'url(' + chrome.extension.getURL('icons/icon_64.png') + ')';
      } else {
        downloads.style.backgroundImage = ''
      }
    });

    return downloads;
}

function setDuration(duration, id) {
  // Falsche und nichtrelevante Videolängen gar nicht erst anzeigen
  if (duration == 0) {
    return;
  }

  var video_length = document.createElement('h5');
  video_length.setAttribute('class', 'duration');
  video_length.textContent = '(' + duration + 'min)';
  $('#downloads_' + id).append(video_length);
}

function createWarning(msg)
{
    var warning = document.createElement('div');
    warning.setAttribute('class', 'warn_text');
    $(warning).html(msg);

    var g1plus_icon = document.createElement('img');
    g1plus_icon.setAttribute('class', 'g1plus_small_icon');
    g1plus_icon.src = chrome.extension.getURL('icons/icon_64.png');
    warning.appendChild(g1plus_icon);

    return warning;
}

/**
 * Erzeugt einen simplen Link.
 *
 * @param url   Url auf die der Link zeigt
 * @param text  Text des Links
 *
 * @return Das Link-Element
 */
function createDownloadLink(url, text)
{
    var downlink = document.createElement('a');
    downlink.setAttribute('href', url);
    downlink.style.display = 'none';
    downlink.textContent = text;

    return downlink;
}

/**
 * Erstellt einen Player, der dem alten GameOne-Player entspricht.
 *
 * @param src  Quelle des anzuzeigenden Videos.
 */
function createLegacyPlayer(src)
{
    var player = document.createElement('embed');
    player.setAttribute('width', 566);
    player.setAttribute('height', 424);
    player.setAttribute('flashvars', 'configParams');
    player.setAttribute('menu', 'false');
    player.setAttribute('swliveconnect', 'true');
    player.setAttribute('allowscriptaccess', 'always');
    player.setAttribute('enablejavascript', 'true');
    player.setAttribute('allowfullscreen', 'true');
    player.setAttribute('quality', 'high');
    player.setAttribute('name', 'embeddedPlayer');
    player.setAttribute('id', 'embeddedPlayer');
    player.setAttribute('src', src);
    player.setAttribute('type', 'application/x-shockwave-flash');
    return player;
}

/**
 * Erstellt einen Player, der dem GameOne-Player entspricht.
 *
 * @param src  Quelle des anzuzeigenden Videos.
 */
function createPlayer(src, parent)
{
    var swf = document.createElement('div');
    parent.appendChild(swf);
    var rand = Math.floor((Math.random()*1000000)+1);
    var flashvars = {
        mrss: src,
        config: "http://www.gameone.de/gameone_de_DE.xml",
        adSite: "gameone.de",
        umaSite: "gameone.de",
        autoPlay: "false",
        url: "",
        tile: "",
        ord: rand,
        image: ""
    };
    var params = {
        wmode: "true",
        enableJavascript: "true",
        allowscriptaccess: "always",
        swLiveConnect: "true",
        allowfullscreen: "true"
    };
    var attributes = {
      id:"embeddedPlayer",
      name:"embeddedPlayer"
    };
    swfobject.embedSWF("http://www.gameone.de/flash/g2player_2.0.60.swf", swf, "566", "424", "9.0.28.0", "expressInstall.swf", flashvars, params, attributes);
}

/**
 * Holt die zugehörigen Downloads des Owner-Objects und fügt sie in einer
 * Download-Box an.
 */
function getDownloads()
{
    try {
        var src = $('#embeddedPlayer', this).get(0).getAttribute('flashvars').split('&').shift();
    } catch(err) {
        var src = $('#embeddedPlayer param[name="flashvars"]', this).val().split('&').shift();
    }
    var id = src.split('-').pop();
    this.appendChild(createDownloadContainer('downloads_' + id));
    node = this;
    $.get('http://gameone.de/api/mrss/' + src, response_mrss(id, node));
}

/**
 * Füllt ein select-Element mit option-Einträgen beginnend bei min bis max
 * aufsteigend. Wird kein callback angegeben entsprech die Einträge den
 * jeweiligen Zahlen.
 *
 * @param select    Das zu befüllende select-Element
 * @param min       Startelement
 * @param max       Endelement
 * @param callback  Funktion die für jedes Element aufgerufen wird. Es wird die
 *                  jeweilige Ziffer übergeben. Rückgabewert muss ein dazu
 *                  korrespondierender String sein.
 */
function addOptions(select, min, max, callback)
{
  for(var i = min; i < max; ++i)
  {
    var option = document.createElement('option');
    if(!callback) {
      option.textContent = i;
    } else {
      option.textContent = callback(i);
    }
    select.appendChild(option);
  }
}

/**
 * Erzeugt eine Altersabfrage, die bei Eingabe eines Alters, welches der
 * Volljährigkeit entspricht, die Freigabe aller altersbeschränkten Inhalte
 * auslöst.
 *
 * @return Das Altersabfrage-Element
 */
function createAgeCheck()
{
  var agecheck = document.createElement('div');
  agecheck.setAttribute('class', 'agecheck g1plus');

  var info = document.createElement('p');
  info.textContent = 'Um Ab-18-Inhalte sehen zu können musst du dein Alter bestätigen:';
  agecheck.appendChild(info);

  var day = document.createElement('select');
  day.setAttribute('class', 'day');
  addOptions(day, 1, 32);
  agecheck.appendChild(day);

  var month = document.createElement('select');
  month.setAttribute('class', 'month');
  addOptions(month, 1, 13);
  agecheck.appendChild(month);

  var year = document.createElement('select');
  year.setAttribute('class', 'year');
  addOptions(year, 0, 100, function(i) {
    var today = new Date();
    return today.getFullYear() - i;
  });
  agecheck.appendChild(year);

  var ok = document.createElement('input');
  ok.setAttribute('type', 'submit');
  ok.setAttribute('value', 'Bestätigen');
  $(ok).click(function() {
    var year = parseInt($('select.year :selected', this.parentNode).text());
    var month = parseInt($('select.month :selected', this.parentNode).text());
    var day = parseInt($('select.day :selected', this.parentNode).text());
    var age = new Date(year, month - 1, day);
    var padded_age = new Date(age.getFullYear() + 18, age.getMonth(), age.getDate());
    var today = new Date();
    if((today.getTime() - padded_age.getTime()) >= 0) {
      var commentable_id = document.getElementById('commentable_id').getAttribute('value');
      $('div.agecheck').empty();
      $('div.agecheck').addClass('loading');
      request_cachedata(commentable_id);
    }
    return false;
  });
  agecheck.appendChild(ok);

  return agecheck;
}

/* Events
 * ==== */

/**
 * Behandeln der Rückgabe der mrss-API. Inhalte die von der flvgen-API geliefert
 * werden, werden gesondert behandelt. Doppelte Urls werden gefiltert.
 */
function response_mrss(id, element)
{
  return function(data) {
    var urls = new Array();
    $(data).filterNode('media:content').each(function() {
      var url = this.getAttribute('url');
      var duration = Math.round(parseFloat(this.getAttribute('duration')) / 60);
      var callback = 'response_mediagen';

      if(url.indexOf('mediaGen.jhtml') != -1) {
        url = 'http://de.esperanto.mtvi.com/www/xml/flv/flvgen.jhtml?vid=' + url.split(':').pop();
        callback = 'response_flvgen';
      } else
        url = url.split('?')[0];

      if(urls.indexOf(url) == -1) {
        urls.push(url);
        setDuration(duration, id);

        if (callback == 'response_mediagen') {
          $.get(url, function(data) { response_mediagen(data, id) } );
        } else {
          $.get(url, function(data) { response_flvgen(data, id) } );
        }
      }
    });

    var preview_image = $(data).filterNode('image').first();

    if (preview_image.length != 0) {
      element.parentNode.style.backgroundImage = 'url(' + encodeURI(preview_image.attr('url')) + ')';
      element.parentNode.style.backgroundRepeat = 'no-repeat';
      element.parentNode.style.backgroundSize = '100%';
    }
  }
}

/**
 * Behandeln von Inhalten, die über die mediagen-API geliefert werden (reguläre
 * Viedos sowie Gametrailers-Videos).
 */
function response_mediagen(data, id)
{
  var downloads = document.getElementById('downloads_' + id);
  var videos = [];

  $('rendition', data).each(function() {
    var v = {};
    v.width = this.getAttribute('width');
    v.height = this.getAttribute('height');
    v.bitrate = this.getAttribute('bitrate');
    v.mime = this.getAttribute('type').split('/').pop();
    if(this.textContent.indexOf('http') == -1) {
      v.url = this.textContent.trim().split('/riptide/').pop();
      v.url = 'http://cdn.riptide-mtvn.com/' + v.url;
    } else {
      v.url = this.textContent;
    }
    videos.push(v);
  });

  videos.sort(function(a, b) {
    return b.width - a.width;
  });

  videos.sort(function(a, b) {
    return b.bitrate - a.bitrate;
  });

  $(videos).each(function () {
    var downlink = createDownloadLink(this.url,
      this.mime + ' ' + this.width + 'x' + this.height + '@' + this.bitrate + 'kbps');
    if (downloads)
      downloads.appendChild(downlink);
  });
}


/**
 * Behandeln von Inhalten, die von der flvgen-API geliefert werden (TV-Folgen 1 - 150).
 */
function response_flvgen(data, id)
{
  var downloads = document.getElementById('downloads_' + id);
  items = [];

  $('src', data).each(function() {
      items.push($(this).text());
  });

  $(items).each(function() {
    var text = this.split('/').pop();
    text = text.split('.').shift();
    var downlink = createDownloadLink(this, text);
    downloads.appendChild(downlink);
  });

  var x = $('a', downloads);

  x.sort(function(a, b) {
    return b.textContent < a.textContent;
  });

  $('a', downloads).each(function() {
    downloads.removeChild(this);
  });

  $(x).each(function() {
    downloads.appendChild(this);
  });
}

/**
 * Wrapper function um this variable aus dem jeweiligen context and die jeweilige Funktion zu übergeben
 * http://stackoverflow.com/questions/939032/jquery-pass-more-parameters-into-callback#answer-939185
 */
function replace_restricted_wrapper(page, index, element) {
  return function(response) {
    if(response.data && response.data[page][String(index + 1)]) {
      var ids = response.data[page][String(index + 1)];
      for(j in ids) {
        console.log('data:' + ids[j]);
        var id = ids[j];
        if(id.indexOf('gallery') > -1) {
            $(element).replaceWith(createWarning('Bei diesem altersbeschränkten Inhalt handelt es sich um eine Bilder-Galerie, Diese werden derzeit nicht von G1Plus erfasst. Dies kann sich in zukünftigen Versionen ändern, wenn gesteigertes Interesse besteht (<a href="https://github.com/g1plus/g1plus/issues/1">Issue #1</a>)'));
        } else {
            var url = "http://www.gameone.de/api/mrss/mgid:gameone:video:mtvnn.com:video_meta-" + id;
            if(id.indexOf('http') > -1) {
                url = id;
            }
            var player_swf = document.createElement('div');
            player_swf.setAttribute('class', 'player_swf');
            var player = createPlayer(url, player_swf);
            $(element).after(player_swf);
            if(id.indexOf('http') == -1) {
                player_swf.getDownloads = getDownloads;
                player_swf.getDownloads(id);
            }
        }
      }
      $(element).remove();
    } else {
      $(element).replaceWith(createWarning('Für diesen altersbeschränkten Inhalt liegt keine Referenz im Cache vor.'));
    }
  }
}

function request_cachedata(commentable_id)
{
  var page = '1';
  var href = window.location.href.split('?').pop().split('/');
  if(href[href.length - 2] == 'part') {
      page = href[href.length - 1];
  }

  // alle gesperreten Inhalte verarbeiten
  $('div.agecheck').each(function(index) {
    // Anfrage an backgroundPage für id
    chrome.extension.sendRequest({func: "getData", id: commentable_id}, replace_restricted_wrapper(page, index, this));
  });
}



/* Main
 * ==== */

// Downloads unter alle Videos einfügen
$('div.player_swf').each(getDownloads);

// altersbeschränkten Inhalte verarbeiten
$('img[src="/images/dummys/dummy_agerated.jpg"]').each(function(i) {
  $(this).replaceWith(createAgeCheck());
});

// load konami code (up up down down left right left right b a enter)
konami = new Konami()
konami.code = function()
{
  $('#header h1').css('background', 'url(http://upload.wikimedia.org/wikipedia/de/thumb/a/a6/GameOneLogo.png/220px-GameOneLogo.png) no-repeat 30px 30px');
  $('#header h1').css('width', '250px');
}
konami.load()